ARCs/arc-0060.md

```diff
@@ -0,0 +1,330 @@
+  data?: any;
+  failingSignData: (StdData | null);
+}
+```
+
+`SignDataError` uses the same error codes of `SignTxnsError` as well as the following codes:
+
+| Status Code | Name | Description |
+| --- | --- | --- |
+| 4600 | Invalid scope | The scope of the signing action is not recognized by the wallet |
+| 4601 | Invalid schema | The schema does not comply with ARC-60 requirements |
+| 4602 | Invalid encoding | The canonical JSON cannot be decoded with the given encoding |
+
+### Semantic Requirements
+
+The call `signData(data, metadata, signer)` **MUST** either return the signed data `ret` or reject the call throwing an error `err`.
+
+> Following [ARC-1](./arc-0001.md) terminology, in this ARC the term **Rejecting** means throwing an error with `err.code=4300`.
+
+Upon calling `signData(data, metadata, signer)`:
+
+- the `data`, `metadata`, and `signer` **MUST NOT** be `null`, otherwise the wallet **MUST** reject the call.
+- `data` **MUST** be validated with respect to the JSON schema `metadata.schema`. If the validation fails, the call **MUST** be rejected with a `4300` error.
+- if the encoding `metadata.encoding` is present, it **MUST** be used to encode `data`.
+- the wallet **MUST** prepare the signing data as `signData=(<data.bytes>)`.
+  - if the value of `ARC6Domain` is not empty than it must used as `prefix`: `signData=(<prefix> || <data.bytes>)`.
+- the wallet **MUST** ask users for signing confirmation. It **MUST** display the `metadata.message` if present, and the `signData` being signed:
+  - if the user approves, `signData` **MUST** be signed with `signer` and `ret` **MUST** be set to the corresponding `SignedDataStr`.
+  - if the user rejects, the call **MUST** fail with error code `4001`.
+
+Note that if `signData` cannot be signed for any reason, the wallet **MUST** throw an error, such that
+
+- `err.message` **SHOULD** indicate the reason of the error (e.g. specify that `data` is not a valid JSON object according to `metadata.schema`)
+- `err.failingSignData` **SHOULD** return the `StdData` object that caused the error, otherwise `null`.
+
+#### Semantic of `data`
+
+- it **MUST** be a valid `StdData` object, otherwise the wallet **MUST** reject.
+- the encoding **MUST** be equal to the value specified with `metadata.encoding` if any, otherwise it **MUST** be UTF-8.
+- if `data` cannot be decoded into a canonicalized JSON object, the wallet **MUST** throw a `4602` error.
+- if the decoded `data` does not comply with the JSON schema in `metadata`, the wallet **MUST** reject.
+- the `ARC60Domain` filed must be validated against forbidden values:
+  - the wallet **MUST** reject if `ARC60Domain="TX"`.
+  - the wallet **MUST** reject if `ARC60Domain="TG"`.
+  - the wallet **MUST** reject if `metadata.scope=MSGSIG` and `ARC60Domain` field is not set to `"arc60"` or `""`.
+- the `bytes` must be a valid byte string:
+  - the wallet **MUST** reject if `bytes` is prepended with a forbidden domain separator `TX` or `TG`.
```

Comment:
Is this validation here independent of the `ARC60Domain` validation? Because if `ARC60Domain` is set to `arc60` (or anything other than `""`, `"TX"`, or `"TG"`), wouldn't it be fine if `bytes` started with `TX` or `TG`? Because the wallet would have to concatenate the value of `ARC60Domain` and `bytes` and hence no malicious payload could be signed?

Reply:
Setting `ARC60Domain` to `""` would make the concatenation not effective. I suggest to always validate `bytes` against malicious payloads (prepended with `TX` or `TG`) to avoid problems.